From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] C2ME: Optimize world gen math

This patch is based on following mixin:
* "com/ishland/c2me/opts/worldgen/vanilla/mixin/structure_weight_sampler/MixinStructureWeightSampler.java"
By: ishland <ishlandmc@yeah.net>
As part of: C2ME-fabric (https://github.com/RelativityMC/C2ME-fabric)
Licensed under: MIT

// SYNC SOURCE START
rev: 22ee25793c20bcfdaa455fd4bbd88c686612cfc3
branch: dev/26.3.0
repo: https://github.com/RelativityMC/C2ME-fabric.git
track:
- "c2me-opts-worldgen-vanilla/src/main/java/com/ishland/c2me/opts/worldgen/vanilla/mixin/structure_weight_sampler/MixinStructureWeightSampler.java"
// SYNC SOURCE END

diff --git a/net/minecraft/world/level/levelgen/Beardifier.java b/net/minecraft/world/level/levelgen/Beardifier.java
index 7041d673e6f6b7b629d325fee770e640f75725a1..7e6a1a0e213c12405ce660a22f4c574199c25f9e 100644
--- a/net/minecraft/world/level/levelgen/Beardifier.java
+++ b/net/minecraft/world/level/levelgen/Beardifier.java
@@ -135,45 +135,52 @@ public class Beardifier implements DensitySampler {
         }
     }
 
+    // Leaf start - C2ME - Optimize world gen math
+    private Beardifier.Rigid[] c2me$pieceArray;
+    private JigsawJunction[] c2me$junctionArray;
+    private void c2me$initArrays() {
+        this.c2me$pieceArray = this.pieces.toArray(Beardifier.Rigid[]::new);
+        this.c2me$junctionArray = this.junctions.toArray(JigsawJunction[]::new);
+    }
+    // Leaf end - C2ME - Optimize world gen math
+
     @Override
     public float sampleValue(final SamplerContext context, final int blockX, final int blockY, final int blockZ) {
         return this.affectedBox != null && this.affectedBox.isInside(blockX, blockY, blockZ) ? this.sampleValueUnchecked(blockX, blockY, blockZ) : 0.0F;
     }
 
     private float sampleValueUnchecked(final int blockX, final int blockY, final int blockZ) {
-        float noiseValue = 0.0F;
-
-        for (Beardifier.Rigid rigid : this.pieces) {
-            BoundingBox box = rigid.box();
-            int groundLevelDelta = rigid.groundLevelDelta();
-            int dx = Math.max(0, Math.max(box.minX() - blockX, blockX - box.maxX()));
-            int dz = Math.max(0, Math.max(box.minZ() - blockZ, blockZ - box.maxZ()));
-            int groundY = box.minY() + groundLevelDelta;
-            int dyToGround = blockY - groundY;
-
-            int dy = switch (rigid.terrainAdjustment()) {
-                case NONE -> 0;
-                case BURY, BEARD_THIN -> dyToGround;
-                case BEARD_BOX -> Math.max(0, Math.max(groundY - blockY, blockY - box.maxY()));
-                case ENCAPSULATE -> Math.max(0, Math.max(box.minY() - blockY, blockY - box.maxY()));
-            };
+        // Leaf start - C2ME - Optimize world gen math
+        if (this.c2me$pieceArray == null || this.c2me$junctionArray == null) {
+            this.c2me$initArrays();
+        }
+
+        float density = 0.0F;
 
-            noiseValue += switch (rigid.terrainAdjustment()) {
+        for (Beardifier.Rigid piece : this.c2me$pieceArray) {
+            BoundingBox blockBox = piece.box();
+            int groundLevelDelta = piece.groundLevelDelta();
+            int distanceXToBox = Math.max(0, Math.max(blockBox.minX() - blockX, blockX - blockBox.maxX()));
+            int distanceZToBox = Math.max(0, Math.max(blockBox.minZ() - blockZ, blockZ - blockBox.maxZ()));
+            int groundY = blockBox.minY() + groundLevelDelta;
+            int distanceToGround = blockY - groundY;
+            density += switch (piece.terrainAdjustment()) { // 2 switch statement merged
                 case NONE -> 0.0F;
-                case BURY -> getBuryContribution(dx, dy / 2.0F, dz);
-                case BEARD_THIN, BEARD_BOX -> getBeardContribution(dx, dy, dz, dyToGround) * 0.8F;
-                case ENCAPSULATE -> getBuryContribution(dx / 2.0F, dy / 2.0F, dz / 2.0F) * 0.8F;
+                case BURY -> getBuryContribution(distanceXToBox, distanceToGround / 2.0F, distanceZToBox);
+                case BEARD_THIN -> getBeardContribution(distanceXToBox, distanceToGround, distanceZToBox, distanceToGround) * 0.8F;
+                case BEARD_BOX -> getBeardContribution(distanceXToBox, Math.max(0, Math.max(groundY - blockY, blockY - blockBox.maxY())), distanceZToBox, distanceToGround) * 0.8F;
+                case ENCAPSULATE -> getBuryContribution(distanceXToBox / 2.0F, Math.max(0, Math.max(blockBox.minY() - blockY, blockY - blockBox.maxY())) / 2.0F, distanceZToBox / 2.0F) * 0.8F;
             };
         }
-
-        for (JigsawJunction junction : this.junctions) {
-            int dx = blockX - junction.getSourceX();
-            int dy = blockY - junction.getSourceGroundY();
-            int dz = blockZ - junction.getSourceZ();
-            noiseValue += getBeardContribution(dx, dy, dz, dy) * 0.4F;
+        for (JigsawJunction jigsawJunction : this.c2me$junctionArray) {
+            int deltaX = blockX - jigsawJunction.getSourceX();
+            int deltaY = blockY - jigsawJunction.getSourceGroundY();
+            int deltaZ = blockZ - jigsawJunction.getSourceZ();
+            density += getBeardContribution(deltaX, deltaY, deltaZ, deltaY) * 0.4F;
         }
 
-        return noiseValue;
+        return density;
+        // Leaf end - C2ME - Optimize world gen math
     }
 
     private static float getBuryContribution(final float dx, final float dy, final float dz) {
